File: 015_LIA/015_LIA.py
'''
Title: Independent Alleles
Rosalind ID: LIA
Rosalind #: 015
URL: http://rosalind.info/problems/lia
'''

import math

def prob_het(k, N):
    ''' 
    Calcula la probabilidad de que al menos N organismos sean AaBb después de k generaciones.
    '''
    prob_AaBb = 4 / 16.0  # Probabilidad de que un organismo sea AaBb

    prob = []
    total = 2 ** k  # Número total de organismos en la generación k
    # Suma de la función de probabilidad binomial general
    for r in range(N, total + 1):
        prob.append(nCr(total, r) * (prob_AaBb ** r) * ((1 - prob_AaBb) ** (total - r)))
    return sum(prob)

def nCr(n, r):
    f = math.factorial
    return f(n) // f(r) // f(n - r)

if __name__ == "__main__":
    with open("./rosalind_lia.txt", 'r') as f:
        k, N = map(int, f.readline().strip().split())
    
    result = prob_het(k, N)
    
    print(f"{result:.6f}")

File: 015_LIA/015_LIA.c
/*
Title: Independent Alleles
Rosalind ID: LIA
Rosalind #: 015
URL: http://rosalind.info/problems/lia
*/
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

long long factorial(int n) {
    long long result = 1;
    for (int i = 2; i <= n; i++) {
        result *= i;
    }
    return result;
}

long long nCr(int n, int r) {
    return factorial(n) / (factorial(r) * factorial(n - r));
}

// Función para calcular la probabilidad
double prob_het(int k, int N) {
    double prob_AaBb = 4.0 / 16.0;
    double prob = 0.0;
    int total = pow(2, k);

    for (int r = N; r <= total; r++) {
        prob += nCr(total, r) * pow(prob_AaBb, r) * pow(1 - prob_AaBb, total - r);
    }

    return prob;
}

int main() {
    FILE *input_file = fopen("rosalind_lia.txt", "r");
    if (input_file == NULL) {
        perror("Error opening file");
        return 1;
    }

    int k, N;
    fscanf(input_file, "%d %d", &k, &N);
    fclose(input_file);

    double result = prob_het(k, N);

    printf("%.6f\n", result);

    return 0;
}